seccomp/umount.go

```diff
@@ -117,15 +124,81 @@ func (u *umountSyscallInfo) process() (*sysResponse, error) {
 	return u.tracer.createContinueResponse(u.reqId), nil
 }
 
+// processUmmountAllowed purpose is to prevent immutable resources from being
+// unmounted.
+//
+// Method will return 'true' if the umount operation is deemed legit, and will
+// return 'false' otherwise.
+func (u *umountSyscallInfo) processUmountAllowed(
+	mip domain.MountInfoParserIface) (bool, *sysResponse) {
+
+	// There must be mountinfo state present for this target. Otherwise, let
+	// kernel handle the error.
+	info := mip.GetInfo(u.Target)
+	if info == nil {
+		return true, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	// Return 'eperm' if the umount operation is over an immutable mountpoint.
+	if u.cntr.IsImmutableMountID(info.MountID) {
+		return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+	}
+
+	// At this point we have already concluded that the mountpoint being
+	// unmounted doesn't match any of the immutable resources, so we could
+	// be tempted to allow all umount instructions reaching this stage to
+	// succeed. However, there is one particular case that we must identify
+	// to prevent the exposure of immutable masked resources. This is the case
+	// of an umount instruction launched from an unshare() context that targets
+	// an immutable resource. The rest of the code in this function deals with
+	// the identification and handling of this particular case.
+
+	p := u.processData
+
+	// Allow umount if this one is launched from a process whose root-inode
```

Comment:
We have a bug here: just because the root is different, it does not mean it's not referencing the same mount. For example, say a process inside the sys container is in a dedicated mount namespace has a root of `/var/lib/`; if that process decides to unmount `/docker`, it's really unmounting `/var/lib/docker` inside the sys container, so it should not be allowed (since `/var/lib/docker` is an immutable mount). But this code will allow it.